gradoop-temporal/src/main/java/org/gradoop/temporal/model/impl/operators/metric/TemporalVertexDegree.java

```diff
@@ -0,0 +1,125 @@
+ * {@link org.gradoop.flink.model.impl.functions.epgm.BySourceId} and/or
+ * {@link org.gradoop.flink.model.impl.functions.epgm.ByTargetId} before you apply this operator.
+ */
+public class TemporalVertexDegree
+  implements UnaryBaseGraphToValueOperator<TemporalGraph, DataSet<Tuple4<GradoopId, Long, Long, Integer>>> {
+
+  /**
+   * The time dimension that will be considered.
+   */
+  private final TimeDimension dimension;
+
+  /**
+   * The degree type (IN, OUT, BOTH);
+   */
+  private final VertexDegree degreeType;
+
+  /**
+   * A flag to decide whether to include the vertex interval as lower and upper bounds for the first and
+   * last degree period. If set to false, the first degree period starts at
+   * {@link org.gradoop.temporal.model.impl.pojo.TemporalElement#DEFAULT_TIME_FROM} and the last degree period
+   * ends at {@link org.gradoop.temporal.model.impl.pojo.TemporalElement#DEFAULT_TIME_TO}.
+   */
+  private boolean includeVertexTime = true;
+
+  /**
+   * Creates an instance of this temporal vertex degree operator.
+   *
+   * @param degreeType the degree type to consider
+   * @param dimension the time dimension to consider
+   */
+  public TemporalVertexDegree(VertexDegree degreeType, TimeDimension dimension) {
+    this.degreeType = Objects.requireNonNull(degreeType);
+    this.dimension = Objects.requireNonNull(dimension);
+  }
+
+  /**
+   * Configure the operator to consider or not the vertex time by calculating the degree evolution.
+   *
+   * @param includeVertexTime true, iff the vertex time is used as lower and upper bounds for the first and
+   *                          last degree period. Note that this requires a join.
+   */
+  public void setIncludeVertexTime(boolean includeVertexTime) {
+    this.includeVertexTime = includeVertexTime;
+  }
+
+  @Override
+  public DataSet<Tuple4<GradoopId, Long, Long, Integer>> execute(TemporalGraph graph) {
+
+    DataSet<Tuple2<GradoopId, TreeMap<Long, Integer>>> edgesWithTrees = graph.getEdges()
+      // 1) Extract vertex id(s) and corresponding time intervals
+      .flatMap(new FlatMapVertexIdEdgeInterval(dimension, degreeType))
+      // 2) Group them by the vertex id
+      .groupBy(0)
+      // 3) For each vertex id, build a degree tree data structure
+      .reduceGroup(new BuildTemporalDegreeTree());
+
+    if (includeVertexTime) {
+      // The time information of the vertex must be included, therefore we need to join.
+      DataSet<Tuple3<GradoopId, Long, Long>> vertexIdInterval = graph.getVertices()
+        .map(new ExtractIdIntervalMap(dimension));
+
+      return edgesWithTrees
+        // 4) Join the vertices to get each vertex interval
+        .join(vertexIdInterval)
+        .where(0).equalTo(0)
+        .with(new ExtendVertexDataWithInterval())
+        // 5) Since join leads to possible divided groups, we need to group again
+        .groupBy(0)
```

Comment:
How is this possible? I don't understand it. GradoopId should be unique in both datasets.

Reply:
Not in a TPGM graph ... (GradoopId + System-Time End timestamp) are the unique constraint